### ISSUE TITLE
Feature request: Command Interpolation

### ISSUE THREAD

**Issue by NotSpaulding** [chrisant996/clink#291]
I would really like Clink to have support for something like bash's "command interpolation"; for example:
```sh
echo `echo Interpolation 1`  # Prints "Interpolation 1"
echo $(echo Interpolation 2) # Prints "Interpolation 2"
```

I am not sure how this would work really or even if Clink would be able to reproduce something similar to this, but it would be a really convenient feature to have.

**Comment by chrisant996** [comment]
Sample script follows.

This script provides simplistic basic command substitution processing.  It does not handle nested substitutions; e.g. `echo $(echo $(echo abc))` will misbehave.  It spawns the commands in new cmd shells; script side effects (e.g. setting env vars) do not affect the original shell.  Etc.

Maybe it's "good enough" for what you're hoping for.

## command_substitution.lua
```lua
local function substitution(line)
    local i = 1
    local result = ''
    while true do
        -- Find a $(command).
        local s,e = line:find('%$%([^)]*%)', i)
        if not s then
            -- Concat the rest of the input line.
            result = result..line:sub(i)
            break
        end

        -- Concat what precedes the $(command).
        result = result..line:sub(i, s - 1)
        i = e + 1

        -- Spawn a new shell to invoke the $(command).
        -- NOTE:  This means the $(command) cannot affect the current shell's
        -- state; for example, it cannot change environment variables.
        local c = line:sub(s + 2, e - 1)
        local f = io.popen(c)
        if f then
            -- Read the command's output.
            local o = f:read('*a') or ''
            
            -- Trim trailing line endings.
            while true do
                local t = o:sub(#o)
                if t ~= '\r' or t ~= '\n' then
                    break
                end
                o = o:sub(1, #o - 1)
            end
            f:close()

            -- Append the output to the input line.
            result = result..o
        end
    end
    return result
end

if clink.onfilterinput then
    clink.onfilterinput(substitution)
end
```

**Comment by NotSpaulding** [comment]
> I think you mean [command substitution](https://www.gnu.org/software/bash/manual/html_node/Command-Substitution.html). Why not the rest of the [shell expansions](https://www.gnu.org/software/bash/manual/html_node/Shell-Expansions.html) as well? And why not the other `$`-related behaviors from bash?

I did not know what the actual name for it was and just came up with something on the spot; the command substitution is all I was after, so I didn't look into anything else, but they still look quite nice.

> The best that Clink could realistically do is always run both `foocommand` and `barcommand`.

That's also what I was thinking, running the "substituted" command in the background, and replacing the user's command with whatever the "substituted" command returns whenever it is done running, temporarily halting the command, maybe even adding a time limit for the "substituted" command to finish

**Comment by NotSpaulding** [comment]
> Sample script follows.
> 
> This script provides simplistic basic command substitution processing. It does not handle nested substitutions; e.g. `echo $(echo $(echo abc))` will misbehave. It spawns the commands in new cmd shells; script side effects (e.g. setting env vars) do not affect the original shell. Etc.
> 
> Maybe it's "good enough" for what you're hoping for.
> ## command_substitution.lua
> 
> ```lua
> --------------------------------------------------------------------------------
> -- Performs simple command substitution similar to bash.
> --
> --  This:
> --      echo $(echo hello)
> --
> --  Emits:
> --      hello
> --
> -- CAVEATS:
> --  - Does not support nested substitutions; neither nested via typing nor
> --    nested via substitution.
> --  - Spawns new cmd shells to invoke commands.  This means commands cannot
> --    affect the current shell's state:  changing env vars or cwd or etc do not
> --    affect the current shell.
> --  - This is a simple implementation, and may not work quite as expected in
> --    some cases.
> 
> local function substitution(line)
>     local i = 1
>     local result = ''
>     local continue
>     while true do
>         -- Find a $(command).
>         local s,e = line:find('%$%([^)]*%)', i)
>         if not s then
>             -- Concat the rest of the input line.
>             result = result..line:sub(i)
>             break
>         end
> 
>         -- Concat what precedes the $(command).
>         result = result..line:sub(i, s - 1)
>         i = e + 1
> 
>         -- Substitution was found, so halt further onfilterinput processing.
>         continue = false
> 
>         -- Spawn a new shell to invoke the $(command).
>         -- NOTE:  This means the $(command) cannot affect the current shell's
>         -- state; for example, it cannot change environment variables.
>         local c = line:sub(s + 2, e - 1)
>         local f = io.popen(c)
>         if f then
>             -- Read the command's output.
>             local o = f:read('*a') or ''
> 
>             -- Trim trailing line endings.
>             while true do
>                 local t = o:sub(#o)
>                 if t ~= '\r' or t ~= '\n' then
>                     break
>                 end
>                 o = o:sub(1, #o - 1)
>             end
>             f:close()
> 
>             -- Append the output to the input line.
>             result = result..o
>         end
>     end
>     return result, continue
> end
> 
> if clink.onfilterinput then
>     clink.onfilterinput(substitution)
> end
> ```

Thanks you so much for this, but I've already stumbled upon an issue:
![](https://user-images.githubusercontent.com/97350911/170099393-e124dd0e-0bb3-44f1-a0d1-4386fc0eec80.png)

the script provided uses the first outputted line, it would be really nice if it could be more like how bash does it, as far as I can tell, bash just replaces newlines with spaces or something similar to that; as shown below:
![](https://user-images.githubusercontent.com/97350911/170099868-053934e0-f776-428b-9c2e-056d48e2b9a0.png)

Again thanks for putting in the time of even making this script in the first place!

**Comment by chrisant996** [comment]
Bash considers newlines to be argument separators.  Cmd considers newlines to be the end of the command.  This is another issue about how in bash this is all built into the shell language syntax, but in cmd it's not.  What you're asking for is either simply not feasible, or is a large amount of work spread over months to chase down all the side effects and ambiguities and quirks and attempt to compensate in ways that approximate similar behavior between bash and cmd.

I shared a sample Lua script to get you started, if you want to try to work through the challenges.  It is not something I intend to invest in:  I see a long list of problems, and I don't see good "general case" solutions for them.  But you may be able to get something working that is "good enough" to meet your specific needs.

**Comment by NotSpaulding** [comment]
Alright I see, again thanks for all of your help, whenever I have time I might build upon the script and share it for other users who might be looking for something like this. Really appreciate all your work!

**Comment by NotSpaulding** [comment]
> Running `barcommand` in that example is highly problematic. Suppose `barcommand` deletes files? This is one of the reasons I will not explore building something like this into Clink -- there is no way to make it work reasonably safely/reliably/accurately.

I don't see how else you'd go about getting the output of a command without running it, I wouldn't be very surprised if bash had a bunch of the issues you can list about implementing this into CMD; the only other approach I can think of is running the command in a sandboxed state which would just lead to more annoyances, less reliability and speed reduction.

> Bash doesn't have time limits. And adding time limits gets hugely complicated and unreliable, and produces unpredictable behavior and unpredictable downstream effects for the rest of the command line, and etc.

I am aware of that, I just think it's an easier and less buggy way of implementing something like this into CMD. I am unsure what bash really does with commands stalling, but I doubt it's something that could be easily implemented into this.

**Comment by chrisant996** [comment]
> > Running `barcommand` in that example is highly problematic. Suppose `barcommand` deletes files? This is one of the reasons I will not explore building something like this into Clink -- there is no way to make it work reasonably safely/reliably/accurately.
> 
> I don't see how else you'd go about getting the output of a command without running it, I wouldn't be very surprised if bash had a bunch of the issues you can list about implementing this into CMD; the only other approach I can think of is running the command in a sandboxed state which would just lead to more annoyances, less reliability and speed reduction.

Maybe the example was unclear.  Let me try again:

**_(Do not run these; they can be destructive, especially if mistyped)_**
- bash:  `if [ 1 = 0 ]; then echo $(echo rm -r .); fi`
- cmd:  `if 1 == 0 echo $(echo rd /s .)`

Bash does not run the `$(echo rm -r .)` because the `if [ 1 = 0 ]` evaluates to false.  Because command substitution is _part of the bash scripting language itself_ and operates at the lowest and most integrated level inside bash.

My point is that Clink would simply be doing dumb string replacements.  And so Clink would run the `$(echo rd /s .)` even though the `if 1 == 0` evaluates to false.

The fact that command substitution is build into bash at the lowest level (**not** as part of the input line editor) is a subtle but profound difference.  I'm not pointing out issues in bash.  I'm saying it's impossible for Clink to behave like bash.  You can sort of approximate it at first glance, but there are major differences.

The idea and request make sense.  But the implications are much broader than they may seem at first glance.

> > Bash doesn't have time limits. And adding time limits gets hugely complicated and unreliable, and produces unpredictable behavior and unpredictable downstream effects for the rest of the command line, and etc.
> 
> I am aware of that, I just think it's an easier and less buggy way of implementing something like this into CMD. I am unsure what bash really does with commands stalling, but I doubt it's something that could be easily implemented into this.

Time limits is not easier or less buggy; it's quite the opposite.

It sounds like there is a supposition that bash has undocumented heuristics to second guess what it was told to do, and unpredictably behave differently.  But I'm pretty sure bash does only precisely what it's told to.  It's the script's responsibility to not tell bash to do something that won't work.  Just like it's a script's responsibility not to tell bash to delete the hard drive by mistake; bash doesn't check whether rm is being asked to do things that might not be desirable under some (or most) circumstances.

**Comment by NotSpaulding** [comment]
> Bash does not run the `$(echo rm -r .)` because the `if [ 1 = 0 ]` evaluates to false. Because command substitution is part of the bash scripting language itself and operates at the lowest and most integrated level inside bash.

I see, it sounds like it might be possible to do something similar, but it would be way too much work to be worth it.

> I'm saying it's impossible for Clink to behave like bash. You can sort of approximate it at first glance, but there are major differences.
> 
> The idea and request make sense, but it's really not possible. The implications are much broader than they may seem at first glance.

You're probably right on this, I wasn't really expecting it to be entirely possible, I am not very familiar with bash to know exactly what could and could not be implemented; just thought it would be something nice to be added. It's very cool that I even got this far into the "implementation", and I really appreciate all the work you've done just for this!

> Time limits is not easier or less buggy; it's quite the opposite.

Oh? Why is this exactly? To me, it sounds like that isn't really the case, an example of a bug I was thinking about that would benefit from a time limit is if it were to never stop, you could program something into the time limit that would terminate the command in a safer way, rather than just pressing `CTRL + C`.

**Comment by chrisant996** [comment]
> > Time limits is not easier or less buggy; it's quite the opposite.
> 
> Oh? Why is this exactly? To me, it sounds like that isn't really the case, an example of a bug I was thinking about that would benefit from a time limit is if it were to never stop, you could program something into the time limit that would terminate the command in a safer way, rather than just pressing `CTRL + C`.

But why would it be a bug for a command to be allowed to run to completion?

I can think of cases where the user specified a command that won't complete -- but that is not a bug, that is user error.  There are also commands that take a long time but are perfectly reasonable or even necessary.

What time limit would be "correct"?

Ask yourself the tough questions.  It should begin to become clear why it creates more problems than it could solve.

**Comment by NotSpaulding** [comment]
That's fair enough, my thought process was just that pressing `CTRL + C` while running a command with substitution would just lead to some really obscure bug that would be difficult to fix, I guess that's not really the case though.
